core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericSummaryParallelIterator.java

```diff
@@ -0,0 +1,355 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import com.google.common.reflect.TypeToken;
+
+import net.opentsdb.data.TimeSeries;
+import net.opentsdb.data.TimeSeriesDataType;
+import net.opentsdb.data.TimeSeriesValue;
+import net.opentsdb.data.TimeStamp;
+import net.opentsdb.data.TypedTimeSeriesIterator;
+import net.opentsdb.data.types.numeric.MutableNumericSummaryValue;
+import net.opentsdb.data.types.numeric.NumericSummaryType;
+import net.opentsdb.data.types.numeric.NumericType;
+import net.opentsdb.exceptions.QueryDownstreamException;
+import net.opentsdb.query.QueryIterator;
+import net.opentsdb.query.QueryNode;
+import net.opentsdb.query.QueryResult;
+import net.opentsdb.query.processor.downsample.DownsampleConfig;
+import net.opentsdb.utils.DateTime;
+
+/**
+ * TODO - longs!
+ * TODO - pluggable aggs.
+ * TODO - interpolators
+ *
+ */
+public class GroupByNumericSummaryParallelIterator implements QueryIterator {
+
+  /** Whether or not NaNs are sentinels or real values. */
+  private final boolean infectious_nan;
+  
+  /** [max_threads][interval]*/
+  private final Accumulator[] accumulators;
+  
+  // TEMP
+  static enum AggEnum {
+    sum, count, min, max, last, avg;
+  }
+  
+  private final int intervals;
+  private final long interval_in_seconds;
+  private final TemporalAmount interval;
+  private final boolean expect_sums_and_counts;
+  private final int start_epoch;
+  private int expected_summary;
+  private final AggEnum agg;
+  private double[] results;
+  private final MutableNumericSummaryValue dp;
+  private int idx;
+  private final TimeStamp ts;
+  
+  /**
+   * Default ctor from a map of time series.
+   * @param node The non-null owner.
+   * @param result A query result to pull the rollup config from.
+   * @param sources A non-null map of sources.
+   */
+  public GroupByNumericSummaryParallelIterator(
+      final QueryNode node, 
+      final QueryResult result,
+      final Map<String, TimeSeries> sources) {
+    this(node, result, sources == null ? null : sources.values());
+  }
+  
+  /**
+   * Alternate ctor with a collection of sources.
+   * @param node The non-null owner.
+   * @param result A query result to pull the rollup config from.
+   * @param sources A non-null collection of sources.
+   */
+  public GroupByNumericSummaryParallelIterator(
+      final QueryNode node, 
+      final QueryResult result,
+      final Collection<TimeSeries> sources) {
+    expected_summary = -1;
+    
+    // TODO
+    expect_sums_and_counts = false;
+    
+    DownsampleConfig downsampleConfig = ((GroupBy) node).getDownsampleConfig();
+    if (null == downsampleConfig) {
+      throw new IllegalStateException("Shouldn't be here if the downsample was null.");
+    }
+    intervals = downsampleConfig.intervals();
+    interval = downsampleConfig.interval();
+    interval_in_seconds = DateTime.parseDuration(downsampleConfig.getInterval()) / 1000;
```

Comment:
We can infer this from downsampleConfig.interval() instead of parsing the interval string again.

Reply:
Can't, tried it out and if you do `interval.get(ChronoUnit.SECONDS)` for a wide time range like 1 month, it throws an unsupported unit exceptions. So we have to double parse it (but eventually we'll move that out.